pkg/yurtmanager/controller/loadbalancerset/loadbalancerset/localbalancerset_controller.go

```diff
@@ -0,0 +1,513 @@
+		configration: c.ComponentConfig.PoolServiceController,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(names.LoadBalancerSetController, mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to PoolService
+	err = c.Watch(&source.Kind{Type: &corev1.Service{}}, &handler.EnqueueRequestForObject{}, NewServicePredicated())
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &netv1alpha1.PoolService{}}, NewPoolServiceEventHandler(), NewPoolServicePredicated())
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &v1beta1.NodePool{}}, NewNodePoolEventHandler(mgr.GetClient()), NewNodePoolPredicated())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups=network.openyurt.io,resources=poolservices,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=network.openyurt.io,resources=poolservices/status,verbs=get;update;patch
+
+// Reconcile reads that state of the cluster for a PoolService object and makes changes based on the state read
+// and what is in the PoolService.Spec
+func (r *ReconcilePoolService) Reconcile(_ context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	// Note !!!!!!!!!!
+	// We strongly recommend use Format() to  encapsulation because Format() can print logs by module
+	// @kadisi
+	klog.Infof(Format("Reconcile PoolService %s/%s", request.Namespace, request.Name))
+
+	service := &corev1.Service{}
+	err := r.Get(context.TODO(), request.NamespacedName, service)
+	if apierrors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to get service %s", request.String())
+	}
+
+	copySvc := service.DeepCopy()
+
+	if needCleanPoolServices(copySvc) {
+		if err := r.reconcileClean(copySvc); err != nil {
+			return reconcile.Result{}, errors.Wrapf(err, "failed to clean pool services")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.reconcileNormal(copySvc); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to reconcile service")
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func needCleanPoolServices(svc *corev1.Service) bool {
+	if !svc.DeletionTimestamp.IsZero() {
+		return true
+	}
+
+	if !isLoadBalancerSetService(svc) {
+		return true
+	}
+
+	return false
+}
+
+func (r *ReconcilePoolService) reconcileClean(svc *corev1.Service) error {
+	currentPoolServices, err := r.currentPoolServices(svc)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get current pool services for service %s/%s", svc.Namespace, svc.Name)
+	}
+
+	if canRemoveFinalizer(currentPoolServices) {
+		if err := r.removeFinalizer(svc); err != nil {
+			return errors.Wrapf(err, "failed to add finalizer")
```

Comment:
failed to add finalizer --> failed to remove finalizer

Reply:
done